File: template/DisjointSet.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-


class DisjointSet:
    '''
    并查集
    '''
    parent = None
    rank = None
    dictMode = False

    def __init__(self, l=0, dictMode=False):
        self.dictMode = dictMode
        # list mode #
        if l > 0:
            self.parent = [n for n in range(l)]
            self.rank = [0] * l
            return
        # dict mode #
        if dictMode:
            self.parent = {}
            self.rank = {}
            return
        raise Exception('ERROR!')

    def findRoot(self, w):
        if self.dictMode and not w in self.parent:
            self.parent[w] = w
            return w
        if self.parent[w] == w:
            return w
        else:
            root = self.findRoot(self.parent[w])
            self.parent[w] = root  # 路径压缩
            return root

    def union(self, w1, w2):
        root1 = self.findRoot(w1)
        root2 = self.findRoot(w2)
        if root1 == root2:
            return
        # dict mode #
        if self.dictMode:
            if not root1 in self.rank:
                self.rank[root1] = 0
            if not root2 in self.rank:
                self.rank[root2] = 0
        if self.rank[root1] < self.rank[root2]:  # rank 合并
            root1, root2 = root2, root1  # swap
        elif self.rank[root1] == self.rank[root2]:
            self.rank[root1] = self.rank[root1] + 1
        self.parent[root2] = root1  # union

    def isEqual(self, x, y):
        return self.findRoot(x) == self.findRoot(y)

    def count(self):
        cnt = 0
        l = len(self.parent)
        for i in range(l):
            if self.parent[i] == i:
                cnt = cnt + 1
        return cnt

File: template/DisjointSet.cpp
#include <bits/stdc++.h>

/**
 * 并查集
 */
class DisjointSet {

private:
    // int parent[MAX];
    int *parent;
    int *rank;
    int len; // len = sizeof(parent) / sizeof(parent[0]);

    // #define MAX 10000
    // struct Node {
    //     int data;
    //     int rank;
    //     int parent;
    // }node[MAX];

public:
    DisjointSet(int n) {
        parent = new int[n];
        rank = new int[n];
        len = n;
        memset(parent, -1, sizeof(int) * n);
        memset(rank, 0, sizeof(int) * n);
    }

    ~DisjointSet() {
        delete[] parent;
        parent = NULL;
        delete[] rank;
        rank = NULL;
        len = 0;
    }

    int findRoot(int n) {
        return parent[n] < 0 ? n : parent[n] = findRoot(parent[n]); // path compression
    }

    int link(int n1, int n2) {
        int root1 = findRoot(n1);
        int root2 = findRoot(n2);
        if (root1 == root2) {
            return root1;
        }
        // rank merge
        if (rank[root1] < rank[root2]) {
            int tmp = root1;
            root1 = root2;
            root2 = tmp;
        } else if (rank[root1] == rank[root2]) {
            ++rank[root1];
        }
        // end rank merge
        // cout << rank[root1] << endl;
        parent[root2] = root1; // union

        return root1;
    }

    bool isEqual(int n1, int n2) {
        return findRoot(n1) == findRoot(n2);
    }

    int count() {
        int cnt = 0;
        for(int i = 0; i < len; ++i) {
            if (parent[i] < 0) {
                ++cnt;
            }
        }
        return cnt;
    }
};
